helix-term/src/handlers/completion.rs

```diff
@@ -291,6 +299,223 @@ fn request_completion(
+                        } else {
+                            "file"
+                        }
+                    }
+                    #[cfg(not(unix))]
+                    "file"
+                };
+
+                let resolved = if path_type == "link" { "resolved " } else { "" };
+
+                let documentation = Some(lsp::Documentation::MarkupContent(lsp::MarkupContent {
+                    kind: lsp::MarkupKind::Markdown,
+                    value: {
+                        #[cfg(unix)]
+                        {
+                            use std::os::unix::prelude::PermissionsExt;
+                            let mode = md.permissions().mode();
+
+                            let perms = [
+                                (libc::S_IRUSR, 'r'),
+                                (libc::S_IWUSR, 'w'),
+                                (libc::S_IXUSR, 'x'),
+                                (libc::S_IRGRP, 'r'),
+                                (libc::S_IWGRP, 'w'),
+                                (libc::S_IXGRP, 'x'),
+                                (libc::S_IROTH, 'r'),
+                                (libc::S_IWOTH, 'w'),
+                                (libc::S_IXOTH, 'x'),
+                            ]
+                            .iter()
+                            .fold(String::new(), |mut acc, (p, s)| {
+                                #[allow(clippy::unnecessary_cast)]
+                                acc.push(if mode & (*p as u32) > 0 { *s } else { '-' });
+                                acc
+                            });
+
+                            // TODO it would be great to be able to individually color the documentation,
+                            // but this will likely require a custom doc implementation (i.e. not `lsp::Documentation`)
+                            // and/or different rendering in completion.rs
+                            format!(
+                                "type: `{path_type}`\n\
+                             permissions: `[{perms}]`\n\
+                             {resolved}full path: `{full_path_name}`",
+                            )
+                        }
+                        #[cfg(not(unix))]
+                        {
+                            format!(
+                                "type: `{path_type}`\n\
+                             {resolved}full path: `{full_path_name}`",
+                            )
+                        }
+                    },
+                }));
+
+                let edit_diff = typed_file_name.as_ref().map(|f| f.len()).unwrap_or(0);
+
+                let text_edit = Some(lsp::CompletionTextEdit::Edit(lsp::TextEdit {
+                    range: range_to_lsp_range(
+                        &text,
+                        Range::new(cursor - edit_diff, cursor),
+                        OffsetEncoding::default(),
+                    ),
+                    new_text: file_name.clone(),
+                }));
```

Comment:
I'd like to try to find a way to avoid digging ourselves into lsp-types here. It makes for a pretty small+focused change but I think we should be introducing core types like `helix_core::Diagnostic` where possible or at least wrappers/replacements like #11486. I suspect it's a potentially large refactor but I think this is the right feature to introduce it with since the actual searching/completion part of this is quite small. (Compared to completing words from the buffer for example)

Reply:
I haven't looked into this much but it might be pretty simple to update the `CompletionItem` struct to have an enum between the `lsp::CompletionItem` and a `helix_core::CompletionItem` struct that has a changeset or transaction instead of text_edits

Reply:
Actually a previous iteration of this PR (e.g. [this](https://github.com/helix-editor/helix/tree/12c088918405b2debd04d259293489f7a1fc539d)) had an enum `CompletionItem` with separating variants `LSP` and `Path` .
I wasn't sure having a separate mechanism additional to `lsp::CompletionItem` would make sense, since there was already the variants `lsp::CompletionItemKind::{FILE, FOLDER}`.

But I guess when thinking about more completion sources it may make sense, especially when the move to more core types resembling those in `lsp` is planned anyways.

So when I'm getting you right, you mean a `CompletionItem` at a lower-level than in `ui::completion` (which I think was the previous iteration). I wonder if it makes sense to generally abstract over `lsp::CompletionItem` and wrap it in `helix_core::CompletionItem`. I should probably investigate this, as I think you're right in that this would need a larger refactor (*maybe* worth it).

I think a `Transaction` would make more sense here, being more general (and we have `impl From<ChangeSet> for Transaction` already)?

Reply:
Ok, I don't think abstracting over `lsp::CompletionItem` makes sense with the current architecture. So rather having a "data" container `helix_core::CompletionItem` similar to `helix_core::Diagnostic` a little bit independent of `lsp::CompletionItem` with data like documentation (possibly lazy) and `Transaction` etc. makes probably more sense for now.

Is it planned btw. to move more logic of this kind from `helix-term` to e.g. `helix-view` or even `helix-core`?
I could imagine `helix_term::ui::CompletionItem` would make more sense in `helix-view`?

Reply:
Yeah generally the more we can move from -term to -view the merrier. Ideally most of what lives in -term right now would be in -view (even commands) but it will take quite a bit of work to decouple the UI elements. So -term would just be the terminal backend and then there could be a -gui crate that would also implement the UI components.

Reply:
Ok, good to know, I'm actually thinking about starting that refactor in this PR, that would obviously blow up this PR a little bit, and would likely produce merge-conflicts with other PRs (e.g. #9996), which is why I'm not sure about it, as I'd like to avoid resolving merge-conflicts and/or maintain this PR for a longer time (in a state where this is more interconnected with the rest of the editor-logic).

Like putting a lot of logic of `helix_term::ui::completion` into e.g. `helix-view::completion` as I don't see many dependencies to `helix-term` currently. (Same for the handlers I guess).

But maybe keep it more incremental, and leave the `CompletionItem` (and the path/lsp `CompletionItem` resolution logic) currently in `helix-term`, i.e. only `helix_term::ui::completion::CompletionItem` to an enum roughly like this:

```rust
pub enum PathKind { Folder, File, Link } // maybe not worth it...

pub enum CompletionItem {
    Lsp {
        item: lsp::CompletionItem,
        id: LanguageServerId,
        resolved: bool,
    },
    Path { // Not sure if this should be a more generic `Core`,
           // but this probably allows being more flexible in the future
        kind: PathKind,
        item: helix_core::CompletionItem,
    },
}

pub struct helix_core::CompletionItem {
    pub transaction: Transaction,
    pub label: String,
    /// Containing Markdown, this could be turned e.g. into an enum to be more structured
    pub documentation: String,
}
```

Reply:
Hmm yeah, working incrementally is probably the right way to go about this. Moving more things to helix-view would mainly help the GUI effort which I don't believe anyone is working on actively anyways. I'm also worried about conflicts with #9801 (though on second look it doesn't look like it would conflict much). An enum like that looks good even if it's only a stepping stone to a bigger refactor later.

Reply:
Actually I have merged the snipped system with path-completion in my personal branch, I don't remember that this caused any serious merge-conflicts...

Ok, good to know, then I'll go the incremental route (when I find the time).